### PR TITLE
PTCD-763 Add Azure Function for removing duplicate provider records from Cosmos

### PR DESCRIPTION
**Pull request by gunndabad**
Also included a read-only 'dry run' function.

### CODE REVIEW

File: src/Dfc.CourseDirectory.Functions/DeleteDuplicateProviderRecords.cs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+            IList<UpdatedProviderInfo> updatedProviders;
+            IList<DeletedProviderInfo> deletedProviders;
+
+            using (var reader = await sqlConnection.QueryMultipleAsync(sql, new { DryRun = dryRun }))
+            {
+                updatedProviders = (await reader.ReadAsync<UpdatedProviderInfo>()).AsList();
+                deletedProviders = (await reader.ReadAsync<DeletedProviderInfo>()).AsList();
+            }
+
+            if (!dryRun)
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Functions/DeleteDuplicateProviderRecords.cs:155]: Personally I'd invert the `if` and early-`return` here to reduce nesting even though it means duplicating the Ok somewhat

File: src/Dfc.CourseDirectory.Functions/DeleteDuplicateProviderRecords.cs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                foreach (var record in updatedProviders)
+                {
+                    await _cosmosDbQueryDispatcher.ExecuteQuery(new UpdateProviderType()
+                    {
+                        ProviderId = record.ProviderId,
+                        ProviderType = record.NewProviderType
+                    });
+                }
+
+                foreach (var record in deletedProviders)
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Functions/DeleteDuplicateProviderRecords.cs:166]: what happens if the sql fails having deleted some, would we then have to work harder to fixup cosmos?

**Review by timabell — APPROVED** [review]

**Review by soharadesign — APPROVED** [review]

**Comment by timabell** [comment]
might be worth wrapping the sql in
```
set xact_abort on
begin tran
...
commit
```

File: src/Dfc.CourseDirectory.Functions/DeleteDuplicateProviderRecords.cs (hunk 2 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Functions/DeleteDuplicateProviderRecords.cs:166]: We could re-run the full sync to get records back into SQL then re-run. Given that we don't have distributed transactions it's never going to be perfect.

**Comment by gunndabad** [comment]
> might be worth wrapping the sql in
> 
> ```
> set xact_abort on
> begin tran
> ...
> commit
> ```

Ah yeh I missed re-adding the transaction.

**Pull request merged**